pymc/sampling/forward.py

```diff
@@ -497,22 +499,85 @@ def sample_posterior_predictive(
 
     Examples
     --------
-    Thin a sampled inferencedata by keeping 1 out of every 5 draws
-    before passing it to sample_posterior_predictive
+    Sample posterior predictive samples any variable included in `var_names`,
+    conditioned on the values of other random variables found in the trace.
+    To ensure the samples are internally consistent, any random variable that depends
+    on another random variable that is being sampled is also sampled.
+    The final list of variables being sampled is shown in the log output.
+
+    Notice that if a random variable has no dependency on other random variables,
+    posterior predictive sampling is equivalent to prior sampling.
+    Likewise, if all random variables are being sampled, the behavior is equivalent to
+    that of sample_prior_predictive.
+
+    Notice how the behavior of the function changes with different `var_names` below:
 
     .. code:: python
 
-        thinned_idata = idata.sel(draw=slice(None, None, 5))
-        with model:
-            idata.extend(pymc.sample_posterior_predictive(thinned_idata))
+        from logging import getLogger
+        import pymc as pm
+
+        # Some environments like google colab suppress the default logging output of PyMC
+        getLogger("pymc").setLevel("INFO")
+
+        kwargs = {"progressbar": False, "random_seed": 0}
+
+        with pm.Model() as m:
+          x = pm.Normal("x")
+          y = pm.Normal("y")
+          z = pm.Normal("z", x + y**2)
+          det = pm.Deterministic("det", pm.math.exp(z))
+          obs = pm.Normal("obs", det, 1, observed=[20])
+
+          idata = pm.sample(tune=10, draws=10, chains=2, **kwargs)
+
+          # Default behavior
+          # Posterior predictive draws of `obs`, conditioned on the posterior draws of `z`
+          pm.sample_posterior_predictive(idata, var_names=["obs"], **kwargs)
+          # Sampling: [obs]
 
-    Generate 5 posterior predictive samples per posterior sample.
+          # Deterministic re-computation of `det` given the posterior draws of `z`
+          pm.sample_posterior_predictive(idata, var_names=["det"], **kwargs)
+          # Sampling: []
+
+          # Posterior predictive draws of `z` and `det`, conditioned on the posterior draws of `x` and `y`
+          # This corresponds to the default behavior if `z` were an observed variable
+          pm.sample_posterior_predictive(idata, var_names=["z", "det"], **kwargs)
+          # Sampling: [z]
+
+          # Posterior predictive draws of `y`, `z` and `det`, conditioned on the posterior draws of `x`
+          # Note that `y` samples are equivalent to the prior, since no other variables influence it.
+          pm.sample_posterior_predictive(idata, var_names=["y", "z", "det"], **kwargs)
```

Comment:
I think it's important to have an example with `y` alone, and the note about priors. Including `z` and `det` here distracts from the point

Reply:
Not sold on it, but change my view

Reply:
The interesting point about this case is that it results in draws from the prior of `y`. Issues related to the interdependence between `z` and `det` are treated in detail below. 

Basically there are two huge gotchas with this function:

1. You sometimes get back priors
2. You have to think about (potentially complex) inter-dependencies between RVs.

(2) is well treated in the examples that follow, but (1) is only given a passing mention. I think it deserves a "full spotlight" example, which could be this one.

Reply:
I am worried about too many examples. I don't think the fact that a posterior predictive sample of a "root" variable corresponds to the prior is particularly relevant. It's just a fun coincidence. I would like more to focus on what happens to z, if you also sample x, y, both or neither. If you understand that logic, the behavior of y should kind of be obvious.

I don't like that people often say that all draws are from the prior, it's just so when you are sampling everything that comes before. Hence my focus on an intermediate variable `z` here.

As a counterpoint, I do mention the prior equivalence three times. Before the example, here as a comment and in the last example where I say it's equivalent to `sample_prior_predictive`

Reply:
OK that sells me on this example as-is. But I don't think anything about this should be assumed obvious ;)

Reply:
I sad conditionally obvious, not obvious!